File: rpg/worldbook_agent.py
"""
worldbook_agent.py — task 84/86: 世界书子代理。

分层信息架构:
  Layer 0: 原文 (script_chapters / document_chunks FTS)
  Layer 1: ChapterFact (chapter_facts) — 每章一行,含 phase/time/locations/characters/events
  Layer 2: PhaseDigest (phase_digests) — 同 phase 多章聚合,含 chapter range + 关键事件
  Layer 3: WorldTimeline — phase_digests 按时间顺序遍列,作为全局时间轴

入口 API:
  consult(script_id, query, *, current_phase=None, current_time=None,
          jump_to_phase=None, jump_to_chapter=None) -> WorldbookResult

返回包含 confidence: 0=完全没匹配, 1.0=精确命中。
GM 拿到 confidence<0.4 应当走"翻阅未果"兜底文案(question op 让玩家确认),
不应硬编一段未知场景。

这是确定性算法, 不调 LLM(快, 低延迟, 可解释)。后续可以加 LLM 重排。

通用性: 任何剧本只要 chapter_facts + phase_digests + worldbook_entries
都灌好, 这个 agent 就能用。不依赖任何特定书的硬编码。
"""
from __future__ import annotations

import re
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Iterator


@dataclass
class WorldbookResult:
    confidence: float = 0.0  # [0, 1]
    timeline_anchor: dict[str, Any] = field(default_factory=dict)  # {phase, chapter, time_label}
    phase_digest: dict[str, Any] | None = None
    chapter_facts: list[dict[str, Any]] = field(default_factory=list)
    worldbook_entries: list[dict[str, Any]] = field(default_factory=list)
    progress_note: str = ""  # 大幅跳跃时的 progress 说明
    sources: list[str] = field(default_factory=list)  # 拉了哪些层
    elapsed_ms: int = 0

    def to_context_text(self) -> str:
        """打包成 GM context bundle 的一段文本。"""
        parts: list[str] = []
        if self.timeline_anchor:
            a = self.timeline_anchor
            parts.append(
                f"=== 当前时间线锚点 ===\n"
                f"故事 phase: {a.get('phase', '(未匹配)')}\n"
                f"参考章节: 第{a.get('chapter_min', '?')}-{a.get('chapter_max', '?')}章\n"
                f"时间标签: {a.get('time_label', '')}"
            )
        if self.phase_digest:
            pd = self.phase_digest
            parts.append(
                f"=== 阶段摘要 ({pd.get('phase_label', '')}) ===\n"
                f"{pd.get('summary', '')[:1500]}"
            )
        if self.chapter_facts:
            lines = []
            for cf in self.chapter_facts[:5]:
                ev = "; ".join(str(e.get("event", "")) for e in (cf.get("events") or [])[:2])
                lines.append(
                    f"第{cf['chapter']}章《{cf.get('title', '')}》｜{cf.get('story_time_label', '')}\n"
                    f"  摘要: {(cf.get('summary') or '')[:200]}\n"
                    f"  事件: {ev[:160]}"
                )
            parts.append("=== 相关章节事实 ===\n" + "\n\n".join(lines))
        if self.worldbook_entries:
            lines = []
            for wb in self.worldbook_entries[:5]:
                lines.append(f"【{wb.get('title', '')}】\n{(wb.get('content') or '')[:400]}")
            parts.append("=== 世界设定 ===\n" + "\n\n".join(lines))
        if self.progress_note:
            parts.append(f"=== 跳跃进度说明 ===\n{self.progress_note}")
        return "\n\n".join(parts)


def consult(
    script_id: int,
    query: str,
    *,
    current_phase: str = "",
    current_time: str = "",
    jump_to_phase: str = "",
    jump_to_chapter: int | None = None,
) -> WorldbookResult:
    """主入口。

    参数:
      script_id      — 当前剧本 id
      query          — 玩家原话 + GM 内部检索关键词
      current_phase  — state.world.timeline.current_phase 当前故事阶段
      current_time   — state.world.time
      jump_to_phase  — 用户大幅时间跳跃的目标 phase
      jump_to_chapter — 或目标章节号

    返回 WorldbookResult, confidence 在 [0, 1]。
    """
    from platform_app.db import connect as _connect

    t0 = time.time()
    result = WorldbookResult()
    if not script_id:
        result.confidence = 0.0
        return result

    try:
        with _connect() as db:
            # 1) 找 timeline anchor: 优先按 jump_to 直接定位, 否则按 current_phase + query 匹配
            anchor = _resolve_anchor(
                db, script_id, query=query,
                current_phase=current_phase,
                current_time=current_time,
                jump_to_phase=jump_to_phase,
                jump_to_chapter=jump_to_chapter,
            )
            if anchor:
                result.timeline_anchor = anchor
                result.sources.append("phase_digests")

            # 2) 拉 PhaseDigest 对应行
            if anchor and anchor.get("phase"):
                row = db.execute(
                    """select phase_label, chapter_min, chapter_max, summary,
                              key_events, key_locations, key_characters,
                              story_time_label_start, story_time_label_end, chapter_count
                       from phase_digests where script_id=%s and phase_label=%s""",
                    (script_id, anchor["phase"]),
                ).fetchone()
                if row:
                    result.phase_digest = dict(row)

            # 3) 拉 相关 ChapterFact
            cmin = anchor.get("chapter_min") if anchor else None
            cmax = anchor.get("chapter_max") if anchor else None
            cf_rows = db.execute(
                """select chapter, title, story_time_label, summary, events
                   from chapter_facts where script_id=%s
                     and (%s::int is null or chapter >= %s)
                     and (%s::int is null or chapter <= %s)
                   order by chapter limit 5""",
                (script_id, cmin, cmin, cmax, cmax),
            ).fetchall()
            result.chapter_facts = [dict(r) for r in cf_rows] if cf_rows else []
            if result.chapter_facts:
                result.sources.append("chapter_facts")

            # 4) Worldbook entries: 高 priority + key 命中
            scan_blob = " ".join([query or "", current_phase or "", current_time or "",
                                  (anchor.get("time_label") if anchor else "") or ""])
            wb_rows = db.execute(
                """select title, content, keys, priority from worldbook_entries
                   where script_id=%s and enabled=true
                   order by priority desc, id asc limit 30""",
                (script_id,),
            ).fetchall()
            picks = []
            for r in wb_rows or []:
                pri = int(r.get("priority") or 50)
                keys = r.get("keys") or []
                hit = pri >= 90 or any(isinstance(k, str) and k and k in scan_blob for k in keys)
                if hit:
                    picks.append(dict(r))
                if len(picks) >= 5:
                    break
            result.worldbook_entries = picks
            if picks:
                result.sources.append("worldbook_entries")

            # 5) 跳跃 progress note
            if jump_to_phase and current_phase and jump_to_phase != current_phase:
                # 拿 jump_to phase 的关键事件给 GM 当 progress
                jp = db.execute(
                    "select phase_label, summary, key_events, chapter_count "
                    "from phase_digests where script_id=%s and phase_label=%s",
                    (script_id, jump_to_phase),
                ).fetchone()
                if jp:
                    events_brief = "; ".join(
                        str(e.get("event", ""))[:80]
                        for e in (jp.get("key_events") or [])[:6]
                        if isinstance(e, dict)
                    )
                    result.progress_note = (
                        f"玩家从 [{current_phase}] 跳到 [{jump_to_phase}]。"
                        f"目标阶段共 {jp.get('chapter_count', '?')} 章,主要进度: {events_brief[:600]}"
                    )

    except Exception as exc:
        result.confidence = 0.0
        result.elapsed_ms = int((time.time() - t0) * 1000)
        result.progress_note = f"(子代理检索异常: {type(exc).__name__})"
        return result

    # 6) 算 confidence
    score = 0.0
    if result.timeline_anchor:
        score += 0.4
    if result.phase_digest:
        score += 0.3
    if result.chapter_facts:
        score += 0.2
    if result.worldbook_entries:
        score += 0.1
    result.confidence = min(1.0, score)
    result.elapsed_ms = int((time.time() - t0) * 1000)
    return result


def _resolve_anchor(
    db, script_id: int, *, query: str, current_phase: str, current_time: str,
    jump_to_phase: str, jump_to_chapter: int | None,
) -> dict[str, Any] | None:
    """决定 timeline anchor。优先级:
    1) jump_to_chapter — 找 chapter 落在哪个 phase
    2) jump_to_phase  — 直接用 phase_label
    3) current_phase  — 用 state 已知 phase
    4) query 关键词 — phase_label 或 time_label 包含 query
    5) fallback: 第一个 phase (chapter 1 起)
    """
    # 1) 章节号
    if jump_to_chapter:
        r = db.execute(
            "select phase_label, chapter_min, chapter_max, story_time_label_start "
            "from phase_digests where script_id=%s and %s between chapter_min and chapter_max "
            "order by chapter_max - chapter_min asc limit 1",
            (script_id, jump_to_chapter),
        ).fetchone()
        if r:
            return _anchor_from_row(r)
    # 2) jump phase
    if jump_to_phase:
        r = db.execute(
            "select phase_label, chapter_min, chapter_max, story_time_label_start "
            "from phase_digests where script_id=%s and phase_label=%s",
            (script_id, jump_to_phase),
        ).fetchone()
        if r:
            return _anchor_from_row(r)
    # 3) current phase
    if current_phase:
        r = db.execute(
            "select phase_label, chapter_min, chapter_max, story_time_label_start "
            "from phase_digests where script_id=%s and phase_label=%s",
            (script_id, current_phase),
        ).fetchone()
        if r:
            return _anchor_from_row(r)
    # 4) query 关键词
    if query:
        terms = [t for t in re.split(r"[\s,，。.\-]+", query) if len(t) >= 2]
        for term in terms[:6]:
            r = db.execute(
                "select phase_label, chapter_min, chapter_max, story_time_label_start "
                "from phase_digests where script_id=%s "
                "  and (phase_label ilike %s or story_time_label_start ilike %s "
                "       or story_time_label_end ilike %s) "
                "order by chapter_min limit 1",
                (script_id, f"%{term}%", f"%{term}%", f"%{term}%"),
            ).fetchone()
            if r:
                return _anchor_from_row(r)
    # 5) fallback: 第一个 phase
    r = db.execute(
        "select phase_label, chapter_min, chapter_max, story_time_label_start "
        "from phase_digests where script_id=%s order by chapter_min asc limit 1",
        (script_id,),
    ).fetchone()
    if r:
        return _anchor_from_row(r)
    return None


def _anchor_from_row(r: dict) -> dict[str, Any]:
    return {
        "phase": r["phase_label"],
        "chapter_min": r["chapter_min"],
        "chapter_max": r["chapter_max"],
        "time_label": r.get("story_time_label_start") or "",
    }


__all__ = ["consult", "WorldbookResult"]

File: rust/crates/rpg-agents/src/worldbook_agent.rs
//! worldbook_agent — 世界书子代理(分层信息架构)。
//!
//! 对应 Python: `rpg/agents/worldbook_agent.py`
//!
//! Layer 0: 原文 (script_chapters / document_chunks FTS)
//! Layer 1: ChapterFact (chapter_facts) — 每章一行
//! Layer 2: PhaseDigest (phase_digests) — 同 phase 多章聚合
//! Layer 3: WorldTimeline — phase_digests 按时间顺序遍列
//!
//! **确定性算法,不调 LLM**(快、低延迟、可解释)。
//!
//! ⚠️ DB 查询全部留 TODO,接 rpg-db 后填。

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sqlx::PgPool;
use std::sync::Arc;

use crate::common::AgentResult;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct WorldbookResult {
    /// [0, 1] 0=完全没匹配, 1.0=精确命中
    pub confidence: f64,
    pub timeline_anchor: Value,
    pub phase_digest: Option<Value>,
    pub chapter_facts: Vec<Value>,
    pub worldbook_entries: Vec<Value>,
    /// 大幅跳跃时的 progress 说明
    pub progress_note: String,
    /// 拉了哪些层
    pub sources: Vec<String>,
    pub elapsed_ms: u64,
}

impl WorldbookResult {
    /// 打包成 GM context bundle 的一段文本。
    pub fn to_context_text(&self) -> String {
        let mut parts: Vec<String> = Vec::new();

        if let Some(a) = self.timeline_anchor.as_object() {
            let phase = a.get("phase").and_then(|v| v.as_str()).unwrap_or("(未匹配)");
            let cmin = a.get("chapter_min").and_then(|v| v.as_str()).unwrap_or("?");
            let cmax = a.get("chapter_max").and_then(|v| v.as_str()).unwrap_or("?");
            let tl = a.get("time_label").and_then(|v| v.as_str()).unwrap_or("");
            parts.push(format!(
                "=== 当前时间线锚点 ===\n故事 phase: {phase}\n参考章节: 第{cmin}-{cmax}章\n时间标签: {tl}"
            ));
        }
        if let Some(pd) = &self.phase_digest {
            let label = pd.get("phase_label").and_then(|v| v.as_str()).unwrap_or("");
            let summary = pd.get("summary").and_then(|v| v.as_str()).unwrap_or("");
            let summary_truncated: String = summary.chars().take(1500).collect();
            parts.push(format!(
                "=== 阶段摘要 ({label}) ===\n{summary_truncated}"
            ));
        }
        if !self.chapter_facts.is_empty() {
            let mut lines: Vec<String> = Vec::new();
            for cf in self.chapter_facts.iter().take(5) {
                let chapter = cf.get("chapter").map(|v| v.to_string()).unwrap_or_default();
                let title = cf.get("title").and_then(|v| v.as_str()).unwrap_or("");
                let time_label = cf
                    .get("story_time_label")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                let summary: String = cf
                    .get("summary")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .chars()
                    .take(200)
                    .collect();
                let events: Vec<String> = cf
                    .get("events")
                    .and_then(|v| v.as_array())
                    .cloned()
                    .unwrap_or_default()
                    .into_iter()
                    .take(2)
                    .map(|e| {
                        e.get("event")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string()
                    })
                    .collect();
                let ev_str: String = events.join("; ").chars().take(160).collect();
                lines.push(format!(
                    "第{chapter}章《{title}》｜{time_label}\n  摘要: {summary}\n  事件: {ev_str}"
                ));
            }
            parts.push(format!("=== 相关章节事实 ===\n{}", lines.join("\n\n")));
        }
        if !self.worldbook_entries.is_empty() {
            let mut lines: Vec<String> = Vec::new();
            for wb in self.worldbook_entries.iter().take(5) {
                let title = wb.get("title").and_then(|v| v.as_str()).unwrap_or("");
                let content: String = wb
                    .get("content")
                    .and_then(|v| v.as_str())
                    .unwrap_or("")
                    .chars()
                    .take(400)
                    .collect();
                lines.push(format!("【{title}】\n{content}"));
            }
            parts.push(format!("=== 世界设定 ===\n{}", lines.join("\n\n")));
        }
        if !self.progress_note.is_empty() {
            parts.push(format!("=== 跳跃进度说明 ===\n{}", self.progress_note));
        }
        parts.join("\n\n")
    }
}

#[derive(Debug, Clone, Default)]
pub struct ConsultInput {
    pub script_id: i64,
    pub query: String,
    pub save_id: Option<i64>,
    pub current_phase: String,
    pub current_time: String,
    pub jump_to_phase: String,
    pub jump_to_chapter: Option<i32>,
}

pub struct WorldbookAgent {
    db: Option<Arc<PgPool>>,
}

impl WorldbookAgent {
    pub fn new() -> Self {
        Self { db: None }
    }

    pub fn with_db(mut self, pool: Arc<PgPool>) -> Self {
        self.db = Some(pool);
        self
    }

    /// 主入口。返回 WorldbookResult。
    ///
    /// 实装路径:
    ///  - Layer 1: 由 anchor 表(script_timeline_anchors)解析 timeline_anchor(若有 db)
    ///  - Layer 2: phase_digests(script-level)按 jump_to_phase 命中
    ///  - Layer 3: chapter_facts 按 jump_to_chapter ±2 取窗口
    ///  - Layer 4: worldbook_entries 按 keyword / aliases 模糊匹配
    ///
    /// db 未注入时降级:仅做 confidence=0,但保留 elapsed。
    pub async fn consult(&self, input: ConsultInput) -> AgentResult<WorldbookResult> {
        let t0 = std::time::Instant::now();
        let mut sources: Vec<String> = Vec::new();
        let mut result = WorldbookResult::default();

        let Some(pool) = self.db.as_ref() else {
            result.confidence = 0.0;
            result.sources.push("db_not_injected".into());
            result.elapsed_ms = t0.elapsed().as_millis() as u64;
            return Ok(result);
        };

        // Layer 1: timeline anchor —— 优先取 jump_to_phase,否则 current_phase。
        let phase_key = if !input.jump_to_phase.is_empty() {
            input.jump_to_phase.clone()
        } else {
            input.current_phase.clone()
        };
        if let Some(anchor) =
            load_timeline_anchor(pool.as_ref(), input.script_id, &phase_key).await
        {
            result.timeline_anchor = anchor;
            sources.push("timeline_anchor".into());
        }

        // Layer 2: phase_digests (剧本级)
        if !phase_key.is_empty() {
            if let Some(pd) = load_phase_digest_by_label(
                pool.as_ref(),
                input.script_id,
                &phase_key,
            )
            .await
            {
                result.phase_digest = Some(pd);
                sources.push("phase_digest".into());
            }
        }

        // Layer 3: chapter_facts 窗口
        if let Some(chapter_min) = input.jump_to_chapter {
            let facts = load_chapter_facts_window(
                pool.as_ref(),
                input.script_id,
                chapter_min - 2,
                chapter_min + 2,
            )
            .await;
            if !facts.is_empty() {
                result.chapter_facts = facts;
                sources.push("chapter_facts".into());
            }
        }

        // Layer 4: worldbook_entries — keyword / aliases 模糊匹配。
        let mut hits = consult_worldbook_entries(
            pool.as_ref(),
            input.script_id,
            input.save_id,
            &input.query,
        )
        .await;
        if !hits.is_empty() {
            // 限 8 个 + 按 priority 倒序(consult_worldbook_entries 内已排)
            hits.truncate(8);
            result.worldbook_entries = hits;
            sources.push("worldbook_entries".into());
        }

        // confidence:有 timeline / digest / chapter_facts / worldbook 任一即基本可用。
        let mut conf = 0.0_f64;
        if result.timeline_anchor.is_object() && !result.timeline_anchor.as_object().unwrap().is_empty() {
            conf += 0.35;
        }
        if result.phase_digest.is_some() {
            conf += 0.25;
        }
        if !result.chapter_facts.is_empty() {
            conf += 0.20;
        }
        if !result.worldbook_entries.is_empty() {
            conf += 0.20;
        }
        result.confidence = conf.min(1.0);
        result.sources = sources;
        result.elapsed_ms = t0.elapsed().as_millis() as u64;
        Ok(result)
    }
}

impl Default for WorldbookAgent {
    fn default() -> Self {
        Self::new()
    }
}

// ── DB 加载辅助 ────────────────────────────────────────────────────────

/// 从 worldbook_entries 表按 keyword / aliases / content 做模糊匹配。
///
/// 返回的 Value 形如 `{"id":..,"title":..,"content":..,"priority":..,"key":..}`,
/// 已按 priority DESC 排序。
async fn consult_worldbook_entries(
    pool: &PgPool,
    script_id: i64,
    save_id: Option<i64>,
    query: &str,
) -> Vec<Value> {
    if query.trim().is_empty() {
        // 仍返回 script 全部,按 priority 排序前 8。
        let entries = rpg_db::repos::worldbook_entries::list_for_script(pool, script_id)
            .await
            .unwrap_or_default();
        return entries
            .into_iter()
            .take(8)
            .map(worldbook_entry_to_value)
            .collect();
    }
    // pattern: 任何子串都触发 ILIKE。
    let pat = format!("%{}%", query.trim().to_lowercase());

    // 自定义 raw query —— key/content/comment/aliases::text ILIKE
    let q = sqlx::query_as::<
        _,
        (
            i64,
            Option<i64>,
            Option<i64>,
            Option<i64>,
            String,
            Value,
            String,
            String,
            bool,
            i32,
            i32,
            Value,
            Value,
        ),
    >(
        r#"SELECT id, script_id, save_id, user_id, key, aliases, content, comment,
                  enabled, priority, token_budget, tags, metadata
           FROM worldbook_entries
           WHERE enabled = true
             AND (script_id = $1 OR save_id = $2)
             AND (
                  LOWER(key) LIKE $3
               OR LOWER(content) LIKE $3
               OR LOWER(comment) LIKE $3
               OR LOWER(COALESCE(aliases::text, '')) LIKE $3
             )
           ORDER BY priority DESC, id ASC
           LIMIT 16"#,
    )
    .bind(script_id)
    .bind(save_id)
    .bind(&pat)
    .fetch_all(pool)
    .await;

    let rows = match q {
        Ok(r) => r,
        Err(e) => {
            tracing::warn!("[worldbook] consult ILIKE 查询失败: {e}");
            return Vec::new();
        }
    };

    rows.into_iter()
        .map(|(id, _sid, _save, _uid, key, aliases, content, comment, _en, priority, _tb, tags, _meta)| {
            json!({
                "id": id,
                "title": key,
                "key": "",
                "content": content,
                "comment": comment,
                "aliases": aliases,
                "priority": priority,
                "tags": tags,
            })
        })
        .collect()
}

fn worldbook_entry_to_value(entry: rpg_db::repos::worldbook_entries::WorldbookEntry) -> Value {
    json!({
        "id": entry.id,
        "title": entry.key,
        "content": entry.content,
        "comment": entry.comment,
        "aliases": entry.aliases,
        "priority": entry.priority,
        "tags": entry.tags,
    })
}

/// 从 script_timeline_anchors 表按 phase 标签加载 anchor。
/// 表不存在 / 行不存在时返回 None。
async fn load_timeline_anchor(pool: &PgPool, script_id: i64, phase_key: &str) -> Option<Value> {
    if phase_key.is_empty() {
        return None;
    }
    let row = sqlx::query_as::<_, (String, Option<i32>, Option<i32>, Option<String>, Option<Value>)>(
        r#"SELECT phase, chapter_min, chapter_max, time_label, metadata
           FROM script_timeline_anchors
           WHERE script_id = $1 AND phase = $2
           LIMIT 1"#,
    )
    .bind(script_id)
    .bind(phase_key)
    .fetch_optional(pool)
    .await
    .ok()
    .flatten();
    row.map(|(phase, cmin, cmax, time_label, metadata)| {
        let cmin_s = cmin.map(|x| x.to_string()).unwrap_or_default();
        let cmax_s = cmax.map(|x| x.to_string()).unwrap_or_default();
        json!({
            "phase": phase,
            "chapter_min": cmin_s,
            "chapter_max": cmax_s,
            "time_label": time_label.unwrap_or_default(),
            "metadata": metadata.unwrap_or(Value::Null),
        })
    })
}

/// 按 phase_label 在 phase_digests(script-level)里命中。
async fn load_phase_digest_by_label(
    pool: &PgPool,
    script_id: i64,
    phase_label: &str,
) -> Option<Value> {
    let row = sqlx::query_as::<_, (i32, String, String, Value)>(
        r#"SELECT phase_index, phase_label, summary, key_events
           FROM phase_digests
           WHERE script_id = $1 AND phase_label = $2
           ORDER BY phase_index DESC
           LIMIT 1"#,
    )
    .bind(script_id)
    .bind(phase_label)
    .fetch_optional(pool)
    .await
    .ok()
    .flatten();
    row.map(|(idx, label, summary, key_events)| {
        json!({
            "phase_index": idx,
            "phase_label": label,
            "summary": summary,
            "key_events": key_events,
        })
    })
}

/// 取 [chapter_min, chapter_max] 窗口内的 chapter_facts。
async fn load_chapter_facts_window(
    pool: &PgPool,
    script_id: i64,
    chapter_min: i32,
    chapter_max: i32,
) -> Vec<Value> {
    let rows = sqlx::query_as::<
        _,
        (
            i32,
            Option<String>,
            Option<String>,
            Option<String>,
            Option<Value>,
        ),
    >(
        r#"SELECT chapter, title, story_time_label, summary, events
           FROM chapter_facts
           WHERE script_id = $1 AND chapter BETWEEN $2 AND $3
           ORDER BY chapter ASC
           LIMIT 20"#,
    )
    .bind(script_id)
    .bind(chapter_min)
    .bind(chapter_max)
    .fetch_all(pool)
    .await
    .unwrap_or_default();
    rows.into_iter()
        .map(|(chapter, title, time_label, summary, events)| {
            json!({
                "chapter": chapter,
                "title": title.unwrap_or_default(),
                "story_time_label": time_label.unwrap_or_default(),
                "summary": summary.unwrap_or_default(),
                "events": events.unwrap_or(Value::Array(vec![])),
            })
        })
        .collect()
}

#[doc(hidden)]
pub async fn load_effective_worldbook_for_save(
    pool: &PgPool,
    script_id: i64,
    save_id: Option<i64>,
) -> AgentResult<Vec<Value>> {
    let mut out: Vec<Value> = Vec::new();
    if let Ok(rows) = rpg_db::repos::worldbook_entries::list_for_script(pool, script_id).await {
        out.extend(rows.into_iter().map(worldbook_entry_to_value));
    }
    if let Some(sid) = save_id {
        if let Ok(rows) = rpg_db::repos::worldbook_entries::list_for_save(pool, sid).await {
            out.extend(rows.into_iter().map(worldbook_entry_to_value));
        }
    }
    Ok(out)
}
